Add support for generating a PI control plane API

sethfowler opened the pull request:
This PR adds support for generating a PI control plane API to `p4c` - in other words, it allows us to generate a PI `P4Info` protocol buffers message that describes the aspects of a P4 program which are exposed to the control plane. That message is serialized to a file in either the usual binary format or (for debugging/testing purposes) in JSON format. This feature is exposed using two new command line options for `p4c`: `--pi-file` requests that the PI API be generated and specifies the output file, and `--pi-as-json`, which requests the JSON serialization format rather than the default binary one.

f491896 turns `libfrontend.a` into a libtool library, `libfrontend.la`. Libtool libraries carry their dependencies with them. Making this change allows `libfrontend` to depend on external libraries (in this case, we'll need the protocol buffers library) without requiring every library and executable that includes `libfrontend` to manually link against them. Unfortunately, this change requires a one-time modification to all `p4c` extensions, but once that's done it'll be easier to make similar changes going forward.

0441bdc adds the dependencies we'll need for serializing PI messages to the `p4c` build system. This means linking in `lprotobuf`, and adding PI as a submodule so we can share the message definitions.

0770e68 is a minor patch to support initializing `ordered_set` with an `initializer_list`, C++11-style. This is used in 86d44ca.

d284df0 is, again, a minor patch. This patch just adds two P4 annotations to the standard metadata struct. `@name` makes us expose it in PI in the same way that it's exposed via the BMV2 JSON -> PI path. `@metadata` just marks it as a metadata struct so that we know to expose it to PI - normally, structs are not exposed directly, although any headers they have as fields are exposed, but we want to fully expose metadata.

86d44ca actually adds the PI control plane API serialization code. This patch is *very* extensively commented, so it will hopefully stand on its own. The steps the code follows are:

- Walk over the program and collect all of the symbols which are exposed to PI.
- Generate a PI id for each symbol. Generating ids requires global knowledge so that, for example, hash collisions are resolved in a stable fashion, so we can only do this once we have all of the symbols.
- Walk over the program again and generate the control plane API, using the ids we generated in the previous step.

The two walks over the program involve some unfortunate code duplicate, though I tried to minimize it, but unfortunately it's not easy to completely get rid of in a clean way. We could get down to a single pass if we changed the rules of PI id generation to accept a little more instability.

It's worth noting that the BMV2 JSON converter and the conversion tool in the PI repo synthesize a number of things that don't exist in the IR - for example, exit tables and `_valid` fields for header types. To maintain compatibility as much as possible, this code does those things too, but we'd be much better off if that sort of thing was either handled in a separate compiler pass or removed altogether. The requirement that we have to support things like header fields that don't exist in the IR has the unfortunate side effect of complicating the code significantly. This is something we can gradually clean up over time.

You'll find a list of the remaining issues at the top of `piSerializer.cpp`. There are still a number of things that aren't supported, but we have a pretty good base here to work from. This code is more complete in some areas than the BMV2 JSON -> PI pathway, but it is missing a few features like direct counters.

sethfowler commented:
I just realized that I forgot to reformat this patch to p4c style (brace placement and such) before uploading it. My mistake. That can be straightened out later; I can't fix it now because I'm doing moving stuff.

mbudiu-vmw commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+  CHECK_NULL(typeMap);
+
+  if (externBlock->type->name == "counter") {
+    serializer.addCounter(externBlock->node->to<IR::IDeclaration>(),
```
Are direct and static counters/meters done the same way?

sethfowler replied:
I had some difficulty understanding how direct counters/meters are supposed to work, so I removed the partial implementation I had for them from this patch. I'll revisit it once I've had a chance to talk them over with someone who understands them better.

mbudiu-vmw commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+
+    // Serialize any header field lists which are passed to standard externs in
+    // the action body.
+    // XXX(seth): This is more hardcoded stuff.
```
Indeed, this is architecture-dependent.
But maybe this could be done in a generic way - at least for extern functions. What should matter is just the function parameter names, not the argument values. In fact, this is just a function implemented by the control-plane called from the data plane. One should not care what the arguments are in the callee.

sethfowler replied:
Handling this better is something that we're actively discussing. We all agree that we need something better than the current approach. =)

mbudiu-vmw commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+
+  auto type = typeMap->getType(declaration->getNode(), true);
+  if (isMetadataType(type)) {
+    // Unlike other top-level structs, top-level metadata structs are fully
```
I don't understand why and how.

sethfowler replied:
Neither do I. This is something we should fix; we should expose structs uniformly to PI. What I've done here to document the current approach, which is extremely hacky.

We were talking about this issue elsewhere; see here:

https://github.com/p4lang/p4c/issues/225#issuecomment-272339779

mbudiu-vmw commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+/*
```
I don't understand whether the goal is to generate a PI which is compatible with the way BMv2 does things today, or this is meant to be a generic PI generation tool. 

How about we sit together for a couple of hours and do by hand the PI generation for a simple P4-16 program? I think that except extern objects everything can be done in a very generic way. Extern objects are trickier, because they are architecture-dependent. Maybe this code can be broken down into architecture-specific and architecture-independent pieces.

sethfowler replied:
We should definitely come up with a better specification for how the table keys work. We can't ignore the types, though, because the keys are not specified using external names, but the header instance fields are. The fact that we have to translate between those two things is the reason that we have to walk the IR. You mention elsewhere that the inliner takes care of updating @name annotations, but I don't think that totally solves the problem because we still need to do with things like indexing header stacks. I would love to be wrong about that, though!

I do want to be clear that the *current* goal is to generate something that matches the existing BMV2 JSON -> PI pathway pretty closely. This makes it much easier to compare the two pathways, which is very handy during development. More importantly, it also makes it feasible for users to migrate from one to the other. Keep in mind that the ids we use at the control plane API level are calculated from the names, so we can't change the names without changing the ids, which will break things for users. Until the BMV2 JSON -> PI pathway is deprecated (which is a place we want to get to ASAP), we've really got to be careful to keep them compatible.

I'd love to talk about this more with you Mihai. We're very much on the same page about making this generic. I'm just trying to get us there incrementally so we can maintain compatibility.

mbudiu-vmw commented:
Actually I have thought more about this overnight. There are really two separate problems:
- generate a PI which can be used by BMv2. In this case we don't have much freedom, our hands are tied by the BMv2 conventions
- generate a "generic" API, which is architecture-independent to the largest degree.

For the first problem I think that the best solution is to write a tool which consumes the JSON that is an input to BMv2. The BMv2 back-end already does a lot of weird conversions, and it is not easily maintainable to have the same conversions be done by the API generation.  

The only problem I know with the BMv2 API is the one with complex table keys. I think that the best way to solve this problem is to actually extend the BMv2 JSON format to take an optional string name for each key field. After all, the expression that you use to feed a key field is not always the best mnemonic for the field itself.

Regarding a generic API generator, this should probably use hierarchical names, following the block inclusion hierarchy (IR::Block). Most names are locally scoped (e.g., an action is local to a table and a table is local to a control), so this makes a lot of sense. I think that going forward with the standard model this is the right way to do things. We should explore a way to specify the extern block APIs, but even in the absence of that we can probably devise a plug-in architecture for the API generation, where each extern block type registers a corresponding class which can generate the API. The only real difficulty here will be making this new style of API interoperate with a BMv2 implementation of the standard model.

ChrisDodd commented on README.md:
````diff
@@ -77,7 +79,7 @@ Installing on macOS:
 - Install dependencies using Homebrew:
   ```
-  brew install autoconf automake libtool bdw-gc boost bison pkg-config
+  brew install autoconf automake libtool bdw-gc boost bison pkg-config protobuf
````
Unfortunately Ubuntu16.04's default protobuf is v2.6.1, which configure.ac apparently thinks is too old, and there's not an easy way to get a newer version on Ubuntu -- need to download the package from github and build?

antoninbas replied:
We use the `proto3` syntax which requires a protobuf version >= 3.0.0. I think it would be nice to mention this in the README and give instructions.

sethfowler replied:
Yes; I wasn't aware that any supported platforms were still shipping protobuf 2.x. I'll add some guidance to `README.md`.

cc10512 commented:
@mbudiu-vmw we shall kickoff the API design group and address these issues in that forum.

sethfowler commented:
Thanks for the thorough review, Mihai. My mistake about the PI repo; it didn't even occur to me that it was private. It's a bit weird to be writing code in a public repo to interface with a tool that only exists in a private repo.

I can't address these issues right now as I'm moving today, but I do want to explicitly address one point: it *is* a goal to have the behavior of this code match the existing BMV2 JSON -> PI path. Once this lands, we can write patches to remove the features we don't want from *both* paths at once (I already have patches for the exit tables). That way, if we find out that we do need those features after all, we can simply revert that patch and both code paths remain in sync. Doing it this way makes it easier to compare the two paths and switch back and forth between them until we're ready to commit fully to this path as the only canonical one.

mbudiu-vmw commented:
If that's the goal I think it's easiest to just ingest the json and forget about the IR.

mbudiu-vmw commented:
I will try to submit a PR which removes the exit tables.

sethfowler commented:
Also, please note that I'm constructing paths manually in part to support annotations like @name, which are not AFAIK represented in any fully qualified name which is automatically tracked by the compiler. This is the notion of "fully qualified external names" I mention several times in the code.

Some of this code would be useful in the larger project, but I wanted to vet it here and then a separate patch can expose it more widely.

sethfowler commented:
Long term we want to remove the BMV2 JSON -> PI path.

Mihai, please don't remove the exit tables before this lands unless you are 1000% sure they are not needed. Otherwise I'd rather remove them from both paths in a separate commit we can revert atomically.

ChrisDodd commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "control-plane/PI"]
+	path = control-plane/PI
+	url = https://github.com/p4lang/PI.git
```
Currently not public, so we should set things up to build without it.  Also, using an https: url rather than a git: one interacts poorly with public key use (particularly important for private repos.)

sethfowler replied:
This should ideally be in a separate repository so there is only one canonical source for the PI protocol buffers definition.

We'll have to put off a decision as to the final approach here. We're looking at opening the PI repo in the near future, so let's wait and see how that goes. I'll update the submodule URL, though.

mbudiu-vmw commented:
The inliner tracks the @name annotations and generates new @name annotations concatenating the instance name with the old @name. The only thing that is missing is the top-level enclosing block name.

mbudiu-vmw commented:
I will submit a PR which removes the exit tables - but you can choose not to merge it right now.
I don't have permissions to update the p4c repo, so I can't mess up anything.
If all tests pass then it is our best proof that these tables are useless.

ChrisDodd commented:
I cherry-picked the unrelated cleanup changes (ordered_set stuff and libtoolize), so you'll need to rebase.

sethfowler commented on Makefile.am:
```diff
@@ -31,6 +33,7 @@ GEN_UNIFIED_CPP=$(TOOLSDIR)/gen-unified-sources.py
 # The first 7 are standard automake variables
 CLEANFILES =                   # Files to remove on clean
 bin_PROGRAMS =                 # Binaries built
+lib_LTLIBRARIES =              # Libtool convenience libraries
```
We should probably actually just get rid of the number here and say something like "these". We have enough maintenance burdens as it is. =)

sethfowler commented on configure.ac:
```diff
@@ -64,6 +64,13 @@ if test "x$CLANG" = "xyes"; then
+AC_SUBST([PROTOBUF_LIBS])
+
+AC_PATH_PROG([PROTOC], [protoc], [])
+AS_IF([test "x$PROTOC" = x], [AC_MSG_ERROR([protoc not found])])
```
Yes, it's part of protobuf. I'll clarify the message.

sethfowler commented on control-plane/Makefile.am:
```diff
@@ -0,0 +1,56 @@
+
+# Work around a bug in protoc v.3.0 which generates unused 'deterministic'
+# parameters that cause a lot of warning spew during the build process. This is
+# fixed in v.3.1, but unfortunately or-tools is not yet compatible.
```
I'll reword the comment to make it a little less surprising.

sethfowler commented on frontends/common/options.cpp:
```diff
@@ -89,6 +89,12 @@ CompilerOptions::CompilerOptions() : Util::Options(defaultMessage) {
+                   "Write a PI control plane API description to the specified file.");
+    registerOption("--pi-as-json", nullptr,
+                   [this](const char* arg) { piAsJson = true; return true; },
+                   "Write out the PI API description as human-readable JSON.");
```
I'm gonna have to push back against this one, Mihai. =) This JSON is explicitly a human-readable representation of the binary format we're using in the PI toolchain itself. JSON is not always easy to read, sure, but it's drastically easier than breaking out a hex editor. =)

It's true that the PI acronym is not well established now, but it will be very shortly once those tools are released to the broader community. I think the existing word is best. If you don't know what PI is, you can't really use the generated file in any case.

antoninbas replied:
Actually I'm curious. If the purpose is to be human-readable, why not use `DebugString` on the proto? Or is it expected that the JSON output may be consumed by other tools?

sethfowler replied:
The JSON output is honestly pretty readable, in my opinion. And you're exactly right: the thought was that using JSON would give us more flexibility to integrate with other tools, testing, etc.

sethfowler commented on p4include/v1model.p4:
```diff
@@ -28,6 +28,7 @@ match_kind {
 }
 
 // Are these correct?
+@metadata @name("standard_metadata")
```
I think we're better off making this data driven. I'm trying to get to the point where we have little or no backend-specific stuff baked into the PI serialization code. (Though we've got a ways to go before we're at that point.)

sethfowler commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+//   header stack instances, which currently requires allowing the programmer to
+//   provide a sequence of ids in the @id pragma. Note that @id *is* supported
+//   for header instance *fields*; this is just about the instances themselves.
+// - Locals are intentionally not exposed, but this prevents table match keys
```
The PI naming scheme reflects the hierarchical structure of the types, which is a subset of taking into account the hierarchical structure of the program. I agree that we should do things as described in the P4-16 spec. We'll gradually get there.

One thing that concerns me is that I may not be using the term "header instance" correctly. This is what I'm referring to do when I say "header instance":

```
header foo_header {
  bit<32> field;
}

struct bar_struct {
  foo_header instance_a;
  foo_header instance_b;
}
```

We must be able to differentiate in the API between `instance_a.field` and `instance_b.field`, of course, and `instance_a` and `instance_b` are the things that I thought were called "header instances". Please let me know if that isn't correct so I can fix the comments and so forth.

mbudiu-vmw replied:
In P4-14 header instance is a very clear concept. So it is in BMv2: they are global variables.
In P4-16 you can have instances of a header type in multiple places. They can be fields in the struct with type H in v1model.p4, or they can be local variables, or they can even be fields of a struct which is used to instantiate local variables anywhere in a local scope.

sethfowler commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+//   are desugared into a match against a local. This could be fixed just by
+//   exposing locals, but first we need to decide how it makes sense to expose
+//   this information to the control plane.
+// - Table match keys which involve header stacks are not supported.
```
They don't need one. It's just an edge case I didn't get to yet.

sethfowler commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+using MatchType = ::p4::config::MatchField::MatchType;
+
+/// @return true if @type has an @metadata annotation.
+static bool isMetadataType(const IR::Type* type) {
```
Sure. I'm not sure what you're getting at here, honestly, Mihai. =) The `@metadata` annotation is just a way to express within P4 the fact that e.g. `standard_metadata` is a special, baked in metadata type. That is, it eliminates the need for us to hardcode `standard_metadata` in this code. We wouldn't need this at all if we were more uniform in the way we exposed struct types to PI, and that's something that I think we should fix, but it will take time and some caution has to be used because it will change ids.

sethfowler replied:
(FWIW, the BMV2 JSON converter also bakes in a second metadata type for "user metadata". I don't really understand what that's for, but `@metadata` allow us to easily support that as well.

mbudiu-vmw replied:
The word "standard_metadata" is part of the P4 v1.0 spec, and it is hardwired in the back-end in several places. It is also available through v1model.h, and reserved (via a comment) in v1model.p4. If your code is specific to BMv2 I don't think there's a need to have a @metadata annotation.

The problem with the annotation is that it needs to be documented someplace, and it also polutes the global annotation namespace. Where is the annotation supposed to be used in P4 programs, and what does it mean? Who sets this annotation?

sethfowler replied:
It currently has some BMv2 things baked in, but the goal is to gradually remove that stuff and make it fully generic.

At this point, this annotation only determines how the type it is applied to is exposed to PI, and it would probably mostly or only be used in target-specific P4 headers. However, we are hardcoding which types are metadata in other areas in the compiler, and we could likely leverage this annotation there as well. Basically, right now we need to synchronize `v1model.p4`, `v1model.h`, and the code that uses `v1model.h` elsewhere in the compiler. It would be ideal to instead express as much of that information as possible directly in P4.

If we have no other uses for `@metadata`, we can phase it out once we phase out the BMv2 JSON -> PI pathway and decide on a consistent representation for struct types in PI. If that's the plan, we could certainly just name it `@_pi_metadata_type` or something like that. I expected `@metadata` to be more broadly useful, but I may be wrong about that. We could also just hardcode `standard_metadata`, but the problem is that that does not cover the "user metadata" that's also included in `v1model.*`, and as far as I can tell, there is no fixed name that identifies the user metadata.

sethfowler commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+
+  const size_t offset;  // This field's offset (really index) within the header.
+  const uint32_t bitwidth;
+  const boost::optional<pi_p4_id_t> id;  // Any '@id' P4 annotation for this field.
```
Please see here:

https://github.com/p4lang/p4c/pull/239#discussion_r96967996

sethfowler commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+/// The global symbols which are exposed to PI. Non-global symbols include
+/// action parameters and header instance fields.
+enum class PISymbolType {
+  ACTION,
```
This is a PI symbol table, and PI does not really have scopes - although the names we generate take into account the hierarchy that exists at the P4 level, the PI namespace is currently flat. The `HeaderFieldPath` and `HeaderField` objects do keep that scope information alive until the field names are serialized into their final PI representation, though.

sethfowler commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+
+  // XXX(seth): These are hardcoded in the BMV2 JSON serializer. Ideally we'd do
+  // something better, but this feature is slated for removal anyway.
+  if (control->name == "ingress" || control->name == "egress") {
```
Agreed. I planned to do it immediately after landing this patch.

sethfowler commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+  CHECK_NULL(typeMap);
+
+  auto parserType = parserBlock->container->type;
+  for (auto param : *parserType->applyParams->getEnumerator()) {
```
This just mirrors what the BMV2 JSON generator does. It makes sense to me to some extent (after all, parsers are the things which populate headers), but I agree that it seems like we should be able to handle this more cleanly. I just need to learn more to understand what the best approach would be.

sethfowler commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+      // that to a ternary match against the '$valid$' field of the appropriate
+      // header instance.
+      // XXX(seth): This is baking in BMV2-specific stuff. We should remove this
+      // by performing the desugaring in a separate compiler pass.
```
That makes sense to me. @antoninbas, what do you think?

antoninbas replied:
Makes sense to me. I believe you would need to update the bmv2 backend to remove the use of the valid match type (https://github.com/p4lang/p4c/blob/master/backends/bmv2/jsonconverter.cpp#L1272).

sethfowler replied:
OK. This is another thing we can probably do fairly quickly after this lands, as long as there's agreement that the name of the boolean field should be `$valid$` across all backends. (I'm not sure whether there might be reason not to do that.)

antoninbas replied:
We can agree on this convention for `isValid()` calls for the time being I believe and use `$valid$ (except if there is a name annotation).

sethfowler replied:
Sounds good. I think I may be missing something about the name annotation part though. =) Is it somehow possible to apply a name annotation to the implicitly validity bit? Or do you mean if there's a name collision?

antoninbas replied:
Well this `$valid$` field is basically needed when a table is using `<hdr>.isValid()` as a match field. Because of the current p4info.proto, match fields have to be listed as `HeaderField` objects. I'm thinking in the longer term, if one gives a name annotation for a match field, one will expect that name to be used in the PI, not `<hdr>.$valid$`. We definitely need to modify p4info.proto to take into account name annotations for match fields. That can be seen as an orthogonal issue to adding `$valid$` to every header type though.

sethfowler replied:
That makes sense! Thanks.

mbudiu-vmw replied:
Removing valid match types may be unnecessary - it's just that the compiler may never generate them. It should probably stay there for backwards compatibility.

sethfowler commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+    // expressions in keys, because such expressions are desugared into a match
+    // against an intermediate local, but before implementing anything we need
+    // to decide how it makes sense to present something like that in PI.
+    if (!path->parent) {
```
It *currently* is, but we can fix it. Let's leave this for now and fix it in a followup PR once we've had a chance to discuss how this should work in more detail.

sethfowler commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+  P4InfoSerializer serializer(symbols, refMap, typeMap);
+
+  // Synthesize a ".drop" action.
+  serializer.addSynthesizedAction(".drop");
```
Totally agreed. That's another thing I intended to do immediately after this PR landed.

sethfowler commented:
Mihai, after reading through this again, I understood better that there's a misunderstanding here I had better clear up. Your idea for generating PI from the BMV2 JSON has already been implemented. That's what I mean by the "BMV2 JSON -> PI" pathway that I keep mentioning. The purpose of this PR is to reimplement that pathway directly in the compiler, in a way that can be gradually refactored and enhanced to be cleaner and more generic.

We need to maintain compatibility with the other pathway if possible in the short term, and that's why this code goes out of its way to reproduce the BMV2 JSON behavior, especially with respect to the generated names, because the names are used to calculate the control plane API ids. Keeping the ids stable is the most important thing to ease the process of migrating away from the "BMV2 JSON -> PI" pathway.

Hopefully this makes it a bit clearer why this code works the way that it does. I'm totally in agreement with you about where we need to go; we just need to go there gradually. =)

sethfowler commented:
Thanks for the review, Antonin. I'll rebase this PR today and hopefully we can get it landed pretty soon.

teverman commented on control-plane/piSerializer.cpp:
```diff
@@ -0,0 +1,1471 @@
+// XXX(seth): Here are the known issues:
+// - Direct counters are not yet supported.
+// - Direct meters are not yet supported.
+// - Action profiles are not yet supported.
```
Action profiles are critical to our use case.  No need to address them in this pull request, but I hope support will be coming soon.

sethfowler replied:
Definitely. We'll get there quickly. Once this PR gets merged, adding more features should be a simpler task.

sethfowler commented:
I've rebased this PR and fixed the various issues pointed out above. Many minor issues were fixed; here are a few:

- Indentation is fixed.
- "PI" is replaced by "P4Runtime" where it makes sense to do so. Some instances have not changed because, for example, the P4Runtime repo is still called "PI".
- Documentation and comments have been improved.
- A note has been added to `README.md` about installing protobuf 3 on Ubuntu.
- The code related to exit tables has been removed since Mihai removed them from the BMV2 JSON converter.

One change which I didn't ultimately make: the submodule URL is still using https rather than ssh. Although the review comments correctly pointed out that there are advantages to using ssh if we want to push directly to the PI repo from the submodule, the disadvantage is that you can't clone the repo at all without having a Github SSH key configured. That affects not only members of the general public, but also Travis CI, so the choice is pretty much made for us. It's still easy to configure a separate push URL that uses ssh, though, so all is not lost.

One other addition: this code triggers a bug on GCC 4.8 which is fixed in more recent versions of GCC and Clang, so I didn't notice until now. (This is exactly why the tests run on GCC 4.8.) Essentially some library code calls `get()` using ADL. The appropriate `get()` overload is chosen by newer compilers with no difficulty, but GCC 4.8 gets confused (I believe because of the involvement of templates, although I'm not sure) and adds the overloaded `get()` functions we have in the global namespace for some of our data structures to the overload resolution list, even though they don't have the right number of arguments and couldn't possibly be selected. It then triggers template instantiations that cause errors. (e.g., it ends up instantiating `Vector<const cstring>` which runs into multiple overloading-related issues that upset the compiler) I've worked around this in 0380b64 by moving the `get()` overloads into a namespace and then importing them back into the global namespace with `using`, which prevents them from being pulled in by ADL. There are other ways to fix this, but IMO the *real* fix is moving off of GCC 4.8 once we're able to.

sethfowler commented:
@ChrisDodd this has been approved by Antonin, but I'd prefer not to merge it without at least a rubber stamp from you. Antonin has the best knowledge of the PI/P4Runtime stuff I'm implementing here, but you're the p4c expert and any feedback you have would be valuable.

sethfowler commented:
Thanks for the review!

Looks like we're good to go except for some cpplint errors. =) I'll fix those up and get this merged.
